### PR TITLE
Implement macro-based deref!() syntax for deref patterns

### PR DESCRIPTION
**Pull request by compiler-errors**
Stop using `box PAT` syntax for deref patterns, and instead use a perma-unstable macro.

Blocked on #122222

r? @Nadrieril

### CODE REVIEW

**Comment by fmease** [comment]
Didn't we want to expand built-in macros to `builtin # ident` going forward?

**Comment by compiler-errors** [comment]
Oh, yeah, that's easier. I didn't even know about that.

File: library/core/src/macros/mod.rs [hunk 1]
```diff
@@ -1714,6 +1714,18 @@ pub(crate) mod builtin {
         /* compiler built-in */
     }
 
+    #[cfg(not(bootstrap))]
+    /// Unstable placeholder for type ascription.
```
Comment (clubby789) [inline library/core/src/macros/mod.rs:1718]: Is this right?

Comment (compiler-errors) [reply library/core/src/macros/mod.rs:1718]: nah lol i am just bad at copying and pasting

File: src/tools/rustfmt/src/patterns.rs
Comment (ytmimi) [inline src/tools/rustfmt/src/patterns.rs:1]: Is `PatKind::Deref(_)` something that could be parsed by the parser and show up in the AST? Just to be safe, can we change all the `unreachable!` calls to just return `None` instead. https://github.com/rust-lang/rustfmt/issues/6105 recently highlighted that a user might pass the expanded output from `-Zunpretty=expanded` to rustfmt and expect that rustfmt won't crash.

**Comment by Nadrieril** [comment]
👍 r=me with the two changes requested by others (when the other PR lands)

File: src/tools/rustfmt/src/patterns.rs
Comment (compiler-errors) [reply src/tools/rustfmt/src/patterns.rs:1]: Yeah, fixed.

**Comment by compiler-errors** [comment]
Good to go I think

@bors r=Nadrieril

**Pull request merged**